read config oidc from vault

### Describe the bug

i'm using quarkus 2.5.1 and when use below config to load oidc config from vault, get error
quarkus.oidc.auth-server-url=${oidc.auth-server-url}

But when i use space in the end of variable problem is fix
quarkus.oidc.auth-server-url=${oidc.auth-server-url: }

what's your idea?


2021-12-06 16:45:36,017 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor#prepareOidcDevConsole threw an exception: java.util.NoSuchElementException: SRCFG00011: Could not expand value oidc.auth-server-url in property quarkus.oidc.auth-server-url
	at io.smallrye.config.ExpressionConfigSourceInterceptor.lambda$getValue$0(ExpressionConfigSourceInterceptor.java:63)
	at io.smallrye.common.expression.ExpressionNode.emit(ExpressionNode.java:22)
	at io.smallrye.common.expression.Expression.evaluateException(Expression.java:56)
	at io.smallrye.common.expression.Expression.evaluate(Expression.java:70)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:56)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:36)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.FallbackConfigSourceInterceptor.getValue(FallbackConfigSourceInterceptor.java:24)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.PropertyNamesConfigSourceInterceptor.getValue(PropertyNamesConfigSourceInterceptor.java:61)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.SmallRyeConfig.getConfigValue(SmallRyeConfig.java:307)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:225)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:167)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.getConfigProperty(OidcDevConsoleProcessor.java:152)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.prepareOidcDevConsole(OidcDevConsoleProcessor.java:78)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:887)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:330)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:252)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:60)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:93)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:450)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:67)
	at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:149)
	at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:105)
	at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:145)
	at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:63)
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor#prepareOidcDevConsole threw an exception: java.util.NoSuchElementException: SRCFG00011: Could not expand value oidc.auth-server-url in property quarkus.oidc.auth-server-url
	at io.smallrye.config.ExpressionConfigSourceInterceptor.lambda$getValue$0(ExpressionConfigSourceInterceptor.java:63)
	at io.smallrye.common.expression.ExpressionNode.emit(ExpressionNode.java:22)
	at io.smallrye.common.expression.Expression.evaluateException(Expression.java:56)
	at io.smallrye.common.expression.Expression.evaluate(Expression.java:70)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:56)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:36)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.FallbackConfigSourceInterceptor.getValue(FallbackConfigSourceInterceptor.java:24)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.PropertyNamesConfigSourceInterceptor.getValue(PropertyNamesConfigSourceInterceptor.java:61)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.SmallRyeConfig.getConfigValue(SmallRyeConfig.java:307)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:225)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:167)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.getConfigProperty(OidcDevConsoleProcessor.java:152)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.prepareOidcDevConsole(OidcDevConsoleProcessor.java:78)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:887)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

	at io.quarkus.builder.Execution.run(Execution.java:116)
	at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:79)
	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:161)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:328)
	... 9 more
Caused by: java.util.NoSuchElementException: SRCFG00011: Could not expand value oidc.auth-server-url in property quarkus.oidc.auth-server-url
	at io.smallrye.config.ExpressionConfigSourceInterceptor.lambda$getValue$0(ExpressionConfigSourceInterceptor.java:63)
	at io.smallrye.common.expression.ExpressionNode.emit(ExpressionNode.java:22)
	at io.smallrye.common.expression.Expression.evaluateException(Expression.java:56)
	at io.smallrye.common.expression.Expression.evaluate(Expression.java:70)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:56)
	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:36)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.FallbackConfigSourceInterceptor.getValue(FallbackConfigSourceInterceptor.java:24)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.PropertyNamesConfigSourceInterceptor.getValue(PropertyNamesConfigSourceInterceptor.java:61)
	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
	at io.smallrye.config.SmallRyeConfig.getConfigValue(SmallRyeConfig.java:307)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:225)
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:167)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.getConfigProperty(OidcDevConsoleProcessor.java:152)
	at io.quarkus.oidc.deployment.devservices.OidcDevConsoleProcessor.prepareOidcDevConsole(OidcDevConsoleProcessor.java:78)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:887)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)



### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

11

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.5.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

mvnw 3.8.1

### Additional information

_No response_

--- a/extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/OidcDevConsoleProcessor.java
+++ b/extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/OidcDevConsoleProcessor.java
@@ -73,7 +73,13 @@ public class OidcDevConsoleProcessor extends AbstractDevConsoleProcessor {
                 closeBuildItem.addCloseTask(closeTask, true);
             }
 
-            String authServerUrl = getConfigProperty(AUTH_SERVER_URL_CONFIG_KEY);
+            String authServerUrl = null;
+            try {
+                authServerUrl = getConfigProperty(AUTH_SERVER_URL_CONFIG_KEY);
+            } catch (Exception ex) {
+                // It is not possible to initialize OIDC Dev Console UI without being able to access this property at the build time
+                return;
+            }
             JsonObject metadata = null;
             if (isDiscoveryEnabled()) {
                 metadata = discoverMetadata(authServerUrl);

--- a/extensions/oidc/deployment/src/main/resources/dev-templates/provider.html
+++ b/extensions/oidc/deployment/src/main/resources/dev-templates/provider.html
@@ -100,13 +100,13 @@ var port = {config:property('quarkus.http.port')};
 
     function signInToOidcProviderAndGetTokens() {
       {#if info:oidcGrantType is 'implicit'}
-        window.location.href = '{info:authorizationUrl}'
+        window.location.href = '{info:authorizationUrl??}'
           + "?client_id=" + '{info:clientId}'
           + "&redirect_uri=" + "http%3A%2F%2Flocalhost%3A" + port + encodedDevRoot + "%2Fio.quarkus.quarkus-oidc%2Fprovider"
           + "&scope=openid&response_type=token id_token&response_mode=query&prompt=login"
           + "&nonce=" + makeid();
       {#else}
-        window.location.href = '{info:authorizationUrl}'
+        window.location.href = '{info:authorizationUrl??}'
           + "?client_id=" + '{info:clientId}'
           + "&redirect_uri=" + "http%3A%2F%2Flocalhost%3A" + port + encodedDevRoot + "%2Fio.quarkus.quarkus-oidc%2Fprovider"
           + "&scope=openid&response_type=code&response_mode=query&prompt=login"
@@ -193,7 +193,7 @@ var port = {config:property('quarkus.http.port')};
     function exchangeCodeForTokens(code){
         $.post("exchangeCodeForTokens",
             {
-              tokenUrl: '{info:tokenUrl}',
+              tokenUrl: '{info:tokenUrl??}',
               client: '{info:clientId}',
               clientSecret: '{info:clientSecret}',
               authorizationCode: code,
@@ -258,7 +258,7 @@ var port = {config:property('quarkus.http.port')};
     function testServiceWithPassword(userName, password, servicePath){
         $.post("testService",
             {
-              tokenUrl: '{info:tokenUrl}',
+              tokenUrl: '{info:tokenUrl??}',
               serviceUrl: "http://localhost:" + port + servicePath,
               client: '{info:clientId}',
               clientSecret: '{info:clientSecret}',
@@ -274,7 +274,7 @@ var port = {config:property('quarkus.http.port')};
     function testServiceWithPasswordInSwaggerUi(userName, password){
         $.post("testService",
             {
-              tokenUrl: '{info:tokenUrl}',
+              tokenUrl: '{info:tokenUrl??}',
               client: '{info:clientId}',
               clientSecret: '{info:clientSecret}',
               user: userName,
@@ -289,7 +289,7 @@ var port = {config:property('quarkus.http.port')};
     function testServiceWithPasswordInGraphQLUi(userName){
         $.post("testService",
             {
-              tokenUrl: '{info:tokenUrl}',
+              tokenUrl: '{info:tokenUrl??}',
               client: '{info:clientId}',
               clientSecret: '{info:clientSecret}',
               user: userName,
@@ -305,7 +305,7 @@ var port = {config:property('quarkus.http.port')};
     function testServiceWithClientCredentials(servicePath) {
         $.post("testService",
             {
-              tokenUrl: '{info:tokenUrl}',
+              tokenUrl: '{info:tokenUrl??}',
               serviceUrl: "http://localhost:" + port + servicePath,
               client: '{info:clientId}',
               clientSecret: '{info:clientSecret}',
@@ -318,7 +318,7 @@ var port = {config:property('quarkus.http.port')};
     function testServiceWithClientCredentialsInSwaggerUi(){
         $.post("testService",
             {
-              tokenUrl: '{info:tokenUrl}',
+              tokenUrl: '{info:tokenUrl??}',
               client: '{info:clientId}',
               clientSecret: '{info:clientSecret}',
               grant: '{info:oidcGrantType}'
@@ -331,7 +331,7 @@ var port = {config:property('quarkus.http.port')};
     function testServiceWithClientCredentialsInGraphQLUi(){
         $.post("testService",
             {
-              tokenUrl: '{info:tokenUrl}',
+              tokenUrl: '{info:tokenUrl??}',
               client: '{info:clientId}',
               clientSecret: '{info:clientSecret}',
               grant: '{info:oidcGrantType}'
@@ -350,8 +350,8 @@ function navigateToSwaggerUiWithToken(token){
         "SecurityScheme":{
             "schema":{
                 "flow":"implicit",
-                "authorizationUrl":"{info:authorizationUrl}",
-                "tokenUrl":"{info:tokenUrl}",
+                "authorizationUrl":"{info:authorizationUrl??}",
+                "tokenUrl":"{info:tokenUrl??}",
                 "type":"oauth2",
                 "description":"Authentication"
             },

--- a/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcConfigPropertySupplier.java
+++ b/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcConfigPropertySupplier.java
@@ -37,21 +37,21 @@ public class OidcConfigPropertySupplier implements Supplier<String> {
         if (defaultValue != null || END_SESSION_PATH_KEY.equals(oidcConfigProperty)) {
             Optional<String> value = ConfigProvider.getConfig().getOptionalValue(oidcConfigProperty, String.class);
             if (value.isPresent()) {
-                return checkUrlProperty(value.get());
+                return checkUrlProperty(value);
             }
             return defaultValue;
         } else {
-            return checkUrlProperty(ConfigProvider.getConfig().getValue(oidcConfigProperty, String.class));
+            return checkUrlProperty(ConfigProvider.getConfig().getOptionalValue(oidcConfigProperty, String.class));
         }
     }
 
-    private String checkUrlProperty(String value) {
-        if (urlProperty && !value.startsWith("http:")) {
-            String authServerUrl = ConfigProvider.getConfig().getValue(AUTH_SERVER_URL_CONFIG_KEY, String.class);
-            return OidcCommonUtils.getOidcEndpointUrl(authServerUrl, Optional.of(value));
-        } else {
-            return value;
+    private String checkUrlProperty(Optional<String> value) {
+        if (urlProperty && value.isPresent() && !value.get().startsWith("http:")) {
+            Optional<String> authServerUrl = ConfigProvider.getConfig().getOptionalValue(AUTH_SERVER_URL_CONFIG_KEY,
+                    String.class);
+            return authServerUrl.isPresent() ? OidcCommonUtils.getOidcEndpointUrl(authServerUrl.get(), value) : null;
         }
+        return value.orElse(null);
     }
 
     public String getOidcConfigProperty() {

--- a/integration-tests/devmode/src/test/java/io/quarkus/test/devconsole/DevConsoleOidcNoDiscoverySmokeTest.java
+++ b/integration-tests/devmode/src/test/java/io/quarkus/test/devconsole/DevConsoleOidcNoDiscoverySmokeTest.java
@@ -1,0 +1,36 @@
+package io.quarkus.test.devconsole;
+
+import org.hamcrest.Matchers;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusDevModeTest;
+import io.restassured.RestAssured;
+
+/**
+ * Note that this test cannot be placed under the relevant {@code -deployment} module because then the DEV UI processor would
+ * not be able to locate the template resources correctly.
+ */
+public class DevConsoleOidcNoDiscoverySmokeTest {
+
+    @RegisterExtension
+    static final QuarkusDevModeTest config = new QuarkusDevModeTest()
+            .withApplicationRoot((jar) -> jar.addAsResource(createApplicationProperties(),
+                    "application.properties"));
+
+    @Test
+    public void testOidcProviderTemplate() {
+        RestAssured.get("q/dev/io.quarkus.quarkus-oidc/provider")
+                .then()
+                .statusCode(200).body(Matchers.containsString("OpenId Connect Dev Console"));
+    }
+
+    private static StringAsset createApplicationProperties() {
+        return new StringAsset("quarkus.oidc.auth-server-url=http://localhost/oidc\n"
+                + "quarkus.oidc.client-id=client\n"
+                + "quarkus.oidc.discovery-enabled=false\n"
+                + "quarkus.oidc.introspection-path=introspect\n");
+
+    }
+}

